src/hyperlink/_url.py

```diff
@@ -1793,46 +1959,61 @@ def replace(self, scheme=_UNSET, host=_UNSET, path=_UNSET, query=_UNSET,
         automatically encoded instead of an error being raised.
         """
         if path is not _UNSET:
-            path = [_encode_reserved(p) for p in path]
+            path = tuple(_encode_reserved(p) for p in path)
         if query is not _UNSET:
-            query = [[_encode_reserved(x)
-                      if x is not None else None
-                      for x in (k, v)]
-                     for k, v in iter_pairs(query)]
+            query = cast(QueryPairs, tuple(
+                tuple(
+                    _encode_reserved(x)
+                    if x is not None else None
+                    for x in (k, v)
+                )
+                for k, v in iter_pairs(query)
+            ))
         if userinfo is not _UNSET:
             if len(userinfo) > 2:
                 raise ValueError('userinfo expected sequence of ["user"] or'
-                                 ' ["user", "password"], got %r' % userinfo)
-            userinfo = u':'.join([_encode_reserved(p) for p in userinfo])
+                                 ' ["user", "password"], got %r' % (userinfo,))
```

Comment:
Here's an example of why `%`-formatting sucks.  If `userinfo` is a tuple, the prior code poops.